daemons/controld/pacemaker-controld.c

```diff
@@ -38,83 +38,74 @@ controld_globals_t controld_globals = {
     .fsa_actions = A_NOTHING,
 };
 
-static pcmk__cli_option_t long_options[] = {
-    // long option, argument type, storage, short option, description, flags
-    {
-        "help", no_argument, NULL, '?',
-        "\tThis text", pcmk__option_default
-    },
-    {
-        "verbose", no_argument, NULL, 'V',
-        "\tIncrease debug output", pcmk__option_default
-    },
-    { 0, 0, 0, 0 }
-};
+static GOptionContext *
+build_arg_context(pcmk__common_args_t *args)
+{
+    return pcmk__build_arg_context(args, NULL, NULL, "[metadata]");
+}
 
 int
 main(int argc, char **argv)
 {
-    int flag;
-    int index = 0;
-    int argerr = 0;
+    crm_exit_t exit_code = CRM_EX_OK;
+    bool initialize = true;
+
     crm_ipc_t *old_instance = NULL;
 
+    GError *error = NULL;
+
+    pcmk__common_args_t *args = pcmk__new_common_args(SUMMARY);
+    gchar **processed_args = pcmk__cmdline_preproc(argv, NULL);
+    GOptionContext *context = build_arg_context(args);
+
     controld_globals.mainloop = g_main_loop_new(NULL, FALSE);
     crm_log_preinit(NULL, argc, argv);
-    pcmk__set_cli_options(NULL, "[options]", long_options,
-                          "daemon for coordinating a Pacemaker cluster's "
-                          "response to events");
-
-    while (1) {
-        flag = pcmk__next_cli_option(argc, argv, &index, NULL);
-        if (flag == -1)
-            break;
-
-        switch (flag) {
-            case 'V':
-                crm_bump_log_level(argc, argv);
-                break;
-            case 'h':          /* Help message */
-                pcmk__cli_help(flag, CRM_EX_OK);
-                break;
-            default:
-                ++argerr;
-                break;
-        }
+
+    if (!g_option_context_parse_strv(context, &processed_args, &error)) {
+        exit_code = CRM_EX_USAGE;
+        goto done;
     }
 
-    if (argc - optind == 1 && pcmk__str_eq("metadata", argv[optind], pcmk__str_casei)) {
+    if (args->version) {
+        g_strfreev(processed_args);
+        pcmk__free_arg_context(context);
+
+        /* FIXME: When pacemaker-attrd is converted to use formatted output,
+         * this can go.
+         */
+        pcmk__cli_help('v', CRM_EX_OK);
+    }
+
+    if ((g_strv_length(processed_args) >= 2)
+        && pcmk__str_eq(processed_args[1], "metadata", pcmk__str_none)) {
         crmd_metadata();
-        return CRM_EX_OK;
-    } else if (argc - optind == 1 && pcmk__str_eq("version", argv[optind], pcmk__str_casei)) {
-        fprintf(stdout, "CRM Version: %s (%s)\n", PACEMAKER_VERSION, BUILD_VERSION);
-        return CRM_EX_OK;
+        initialize = false;
+        goto done;
     }
 
     crm_log_init(NULL, LOG_INFO, TRUE, FALSE, argc, argv, FALSE);
 
-    if (optind > argc) {
-        ++argerr;
-    }
-
-    if (argerr) {
-        pcmk__cli_help('?', CRM_EX_USAGE);
+    for (int i = 0; i < args->verbosity; i++) {
+        crm_bump_log_level(argc, argv);
```

Comment:
You can use `pcmk__cli_init_logging("pacemaker-controld", args->verbosity)` instead of this block.

Reply:
Thanks, I was confused about that. When I was trying to figure it out, I came across comments in `cts-{fence,exec}-helper.c` that say
```
/* We have to use crm_log_init here to set up the logging because there's
     * different handling for daemons vs. command line programs, and
     * pcmk__cli_init_logging is set up to only handle the latter.
     */
```

Reply:
And right after that:
```
crm_log_init(NULL, LOG_INFO, TRUE, (verbose? TRUE : FALSE), argc, argv,
                 FALSE);

    for (int i = 0; i < args->verbosity; i++) {
        crm_bump_log_level(argc, argv);
    }
```

Reply:
Right, the options are different for daemons and tools, so we can't just use pcmk__cli_init_logging() unconditionally. For comparison, see pacemakerd's use of it for non-daemonizing options only. Once everything's converted a more streamlined solution might become apparent.